probe-rs/src/probe/blackmagic/mod.rs

```diff
@@ -0,0 +1,1523 @@
+                if *tdi { '1' } else { '0' }
+            ),
+            RemoteCommand::JtagInit => write!(f, "+#!JS#"),
+            RemoteCommand::JtagReset => write!(f, "+#!JR#"),
+            RemoteCommand::JtagTms { bits, length } => {
+                write!(f, "!JT{:02x}{:x}#", *length, *bits)
+            }
+            RemoteCommand::JtagTdi { bits, length, tms } => write!(
+                f,
+                "!J{}{:02x}{:x}#",
+                if *tms { 'D' } else { 'd' },
+                *length,
+                *bits
+            ),
+            RemoteCommand::JtagAddDevice {
+                index,
+                dr_prescan,
+                dr_postscan,
+                ir_len,
+                ir_prescan,
+                ir_postscan,
+                current_ir,
+            } => write!(
+                f,
+                "!HJ{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:08x}#",
+                *index, *dr_prescan, *dr_postscan, *ir_len, *ir_prescan, *ir_postscan, *current_ir
+            ),
+            RemoteCommand::SwdInit => write!(f, "!SS#"),
+            RemoteCommand::SwdIn { length: bits } => {
+                write!(f, "!Si{:02x}#", *bits)
+            }
+            RemoteCommand::SwdInParity { length } => {
+                write!(f, "!SI{:02x}#", *length)
+            }
+            RemoteCommand::SwdOut { value, length } => {
+                write!(f, "!So{:02x}{:x}#", *length, *value)
+            }
+            RemoteCommand::SwdOutParity { value, length } => {
+                write!(f, "!SO{:02x}{:x}#", *length, *value)
+            }
+            RemoteCommand::TargetReset(reset) => {
+                write!(f, "!GZ{}#", if *reset { '1' } else { '0' })
+            }
+            RemoteCommand::GetAccelerators => {
+                write!(f, "!HA#")
+            }
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum RemoteError {
+    ParameterError(u64),
+    Error(u64),
+    Unsupported(u64),
+    ProbeError(std::io::Error),
+}
+
+impl core::fmt::Display for RemoteError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::ParameterError(e) => write!(f, "Remote paramater error with result {:016x}", *e),
+            Self::Error(e) => write!(f, "Remote error with result {:016x}", *e),
+            Self::Unsupported(e) => write!(f, "Remote command unsupported with result {:016x}", *e),
+            Self::ProbeError(e) => write!(f, "Probe error {}", e),
+        }
+    }
+}
+
+impl ProbeError for RemoteError {}
+
+struct RemoteResponse(u64);
+
+#[derive(PartialEq, Copy, Clone)]
+enum SwdDirection {
+    Input,
+    Output,
+}
+
+impl From<bool> for SwdDirection {
+    fn from(value: bool) -> Self {
+        if value {
+            SwdDirection::Output
+        } else {
+            SwdDirection::Input
+        }
+    }
+}
+
+/// A Black Magic Probe.
+pub struct BlackMagicProbe<R: Read, W: Write> {
```

Comment:
Can we use trait objects instead of the generics here?

Reply:
Good idea. Again, the concept of using heap-allocation is still somewhat novel. I've reworked it to use boxed traits, and the impl statements got much simpler.